crates/bevy_ecs/src/reflect/component.rs

```diff
@@ -290,3 +284,19 @@ impl<C: Component + Reflect + FromWorld> FromType<C> for ReflectComponent {
         })
     }
 }
+
+fn from_reflect_or_world<T: FromReflect>(reflected: &dyn Reflect, world: &mut World) -> T {
```

Comment:
This should have some doc strings and a test: this fallback behavior is quite subtle.